pkg/apiserver/slowquery/queries.go

```diff
@@ -106,46 +103,79 @@ type GetListRequest struct {
 	// for showing slow queries in the statement detail page
 	Plans  []string `json:"plans" form:"plans"`
 	Digest string   `json:"digest" form:"digest"`
-}
 
-type GetDetailRequest struct {
-	Digest    string  `json:"digest" form:"digest"`
-	Time      float64 `json:"time" form:"time"`
-	ConnectID int64   `json:"connect_id" form:"connect_id"`
+	Fields string `json:"fields" form:"fields"` // example: "Query,Digest"
 }
 
-func getAllColumnNames() []string {
-	t := reflect.TypeOf(Base{})
+func getRefColumns(jsonFields ...string) ([]string, error) {
+	fields := make(map[string]*reflect.StructField)
+	t := reflect.TypeOf(SlowQuery{})
 	fieldsNum := t.NumField()
-	ret := make([]string, 0, fieldsNum)
 	for i := 0; i < fieldsNum; i++ {
 		field := t.Field(i)
-		if s, ok := field.Tag.Lookup("gorm"); ok {
-			list := strings.Split(s, ":")
-			if len(list) < 1 {
-				panic(fmt.Sprintf("Unknown gorm tag field: %s", s))
+		fields[strings.ToLower(field.Tag.Get("json"))] = &field
+	}
+	ret := make([]string, 0, len(jsonFields))
+	for _, fieldName := range jsonFields {
+		if field, ok := fields[strings.ToLower(fieldName)]; ok {
+			if convert, ok := field.Tag.Lookup("convert"); ok {
+				ret = append(ret, fmt.Sprintf("%s AS %s", convert, gorm.ToColumnName(field.Name)))
+			} else if s, ok := field.Tag.Lookup("gorm"); ok {
+				list := strings.Split(s, ":")
+				if len(list) < 1 {
+					return nil, fmt.Errorf("unknown gorm tag field: %s", s)
+				}
+				ret = append(ret, list[1])
+			} else {
+				return nil, fmt.Errorf("field %s cannot find ref column", fieldName)
 			}
-			ret = append(ret, list[1])
+		} else {
+			return nil, fmt.Errorf("unknown field %s", fieldName)
 		}
 	}
-	ret = append(ret, "Time")
-	return ret
+	return ret, nil
 }
 
-func isValidColumnName(name string) bool {
-	for _, item := range getAllColumnNames() {
-		if name == item {
-			return true
+func getRefColumn(jsonField string) (string, error) {
+	t := reflect.TypeOf(SlowQuery{})
+	fieldsNum := t.NumField()
+	for i := 0; i < fieldsNum; i++ {
+		field := t.Field(i)
+		if field.Tag.Get("json") == jsonField {
+			if s, ok := field.Tag.Lookup("gorm"); ok {
+				list := strings.Split(s, ":")
+				if len(list) < 1 {
+					return "", fmt.Errorf("unknown gorm tag field: %s", s)
+				}
+				return list[1], nil
+			}
+			return "", fmt.Errorf("field %s cannot find ref column", jsonField)
 		}
 	}
-	return false
+	return "", nil
+}
+
+type GetDetailRequest struct {
+	Digest    string  `json:"digest" form:"digest"`
+	Timestamp float64 `json:"timestamp" form:"timestamp"`
+	ConnectID int64   `json:"connect_id" form:"connect_id"`
 }
 
-func QuerySlowLogList(db *gorm.DB, req *GetListRequest) ([]Base, error) {
+func QuerySlowLogList(db *gorm.DB, req *GetListRequest) ([]SlowQuery, error) {
+	refColumns, _ := getRefColumns("digest", "connection_id", "timestamp") // "digest", "connection_id", "timestamp" for detail
```

Comment:
How about:

```go
columnNames := strings.Split(strings.TrimSpace(req.Fields), ",")
refColumns := getRefColumns("digest", "connection_id", "timestamp", columnNames...)
```

Reply:
By the way the split can be avoided (see https://github.com/pingcap-incubator/tidb-dashboard/pull/749#discussion_r487866679)

Reply:
It prompts compile error:

![image](https://user-images.githubusercontent.com/1284531/93315827-1445c900-f83e-11ea-8202-cf17bda71e62.png)

Reply:
Change it to:

```go
	sqlFields := []string{"digest", "connection_id", "timestamp"}
	if strings.TrimSpace(req.Fields) != "" {
		sqlFields = append(sqlFields, strings.Split(req.Fields, ",")...)
		sqlFields = funk.UniqString(sqlFields)
	}
	refColumns, err := getRefColumns(sqlFields...)
```